### ISSUE TITLE
LLM Integration

### ISSUE THREAD

**Issue by liamca** [microsoft/markitdown#12]
As a feature suggestion, I would love if this could allow you to plug in an LLM (such as GPT-4o) where images that are included in the content could be sent to the LLM for further understanding. For example, if it was say a bar chart or graph, it would be sent to convert that to a Markdown table and be integrated in the final Markdown output.

**Comment by dolbex** [comment]
I could see something coming back giving us markers of where that content _should_ appear along with the source content in a way that we could handle it as we please. Ex: Send to LLM, upload to a server and link, etc.

**Comment by janzheng** [comment]
maybe as an MCP server? That wouldn't be too hard

**Comment by gagb** [comment]
Relevant: https://github.com/microsoft/markitdown/blob/main/src/markitdown/_markitdown.py#L755 
Ive tested an older version of it, but not the new version.

**Comment by sdhutchins** [comment]
@gagb Would be great to have this as an example in the README! Thanks.

**Comment by liamca** [comment]
> [@gagb](https://github.com/gagb) Would be great to have this as an example in the README! Thanks.

Agreed. IMO, a PDF based example would be best where it is able to do the text in the normal way, but any images in the PDF are sent to the LLM.  I think this is a more compelling example to organizations then just PNG's.

**Comment by alejlatorre** [comment]
> > As a feature suggestion, I would love if this could allow you to plug in an LLM (such as GPT-4o) where images that are included in the content could be sent to the LLM for further understanding. For example, if it was say a bar chart or graph, it would be sent to convert that to a Markdown table and be integrated in the final Markdown output.
> 
> It's already supported, here you go. You need an openai key.
> 
> from markitdown import MarkItDown
> from openai import OpenAI
> 
> client = OpenAI()
> md = MarkItDown(mlm_client=client, mlm_model="gpt-4o")
> image_file = 'https://gagb.github.io/imgs/bansal-chi21.png'
> result = md.convert(source=image_file)
> print(result.text_content)
> Output
> 
> ImageSize: 388x368
> 
> # Description:
> This image illustrates a comparison of decision accuracy with and without explanations. On the left, there's a computer icon connected to a user symbol, suggesting human-computer interaction. The vertical axis represents the "Accuracy of decisions," ranging from 0.0 to 1.0.
> 
> Two dotted lines extend horizontally: one labeled "R only" and another higher up labeled "R+Explanation." A red arrow points upward from "R only" to "R+Explanation," indicating an increase in accuracy with the addition of explanations. The red delta symbol (∆) signifies the improvement, which is greater than zero. A blue question mark next to it suggests uncertainty or inquiry related to this improvement.
> 
> At the bottom, the label "B Ours" signifies that this concept is part of the author's work or findings. Overall, the diagram emphasizes the value of explanations in enhancing decision-making accuracy in human-computer interactions.

Hi! Maybe I did not find the way and is already implemented but that looks more like the description of the image rather than the text extraction for example. Is there any way to customize the prompt?

**Comment by Soulter** [comment]
> > > As a feature suggestion, I would love if this could allow you to plug in an LLM (such as GPT-4o) where images that are included in the content could be sent to the LLM for further understanding. For example, if it was say a bar chart or graph, it would be sent to convert that to a Markdown table and be integrated in the final Markdown output.
> > 
> > 
> > It's already supported, here you go. You need an openai key.
> > from markitdown import MarkItDown
> > from openai import OpenAI
> > client = OpenAI()
> > md = MarkItDown(mlm_client=client, mlm_model="gpt-4o")
> > image_file = 'https://gagb.github.io/imgs/bansal-chi21.png'
> > result = md.convert(source=image_file)
> > print(result.text_content)
> > Output
> > ImageSize: 388x368
> > # Description:
> > This image illustrates a comparison of decision accuracy with and without explanations. On the left, there's a computer icon connected to a user symbol, suggesting human-computer interaction. The vertical axis represents the "Accuracy of decisions," ranging from 0.0 to 1.0.
> > Two dotted lines extend horizontally: one labeled "R only" and another higher up labeled "R+Explanation." A red arrow points upward from "R only" to "R+Explanation," indicating an increase in accuracy with the addition of explanations. The red delta symbol (∆) signifies the improvement, which is greater than zero. A blue question mark next to it suggests uncertainty or inquiry related to this improvement.
> > At the bottom, the label "B Ours" signifies that this concept is part of the author's work or findings. Overall, the diagram emphasizes the value of explanations in enhancing decision-making accuracy in human-computer interactions.
> 
> Hi! Maybe I did not find the way and is already implemented but that looks more like the description of the image rather than the text extraction for example. Is there any way to customize the prompt?

It can customize prompt, just pass the `mlm_prompt` to `convert` method. 

Like this:

```py
# ...
prompt = "extract the objects in the picture, and only output keywords, separated by commas."
result = markitdown.convert("path/to/example.jpg", mlm_prompt=prompt)
```

But the latest version cannot apply the param properly when trying to convert an url file, and I just created a pr to fix it. #48

**Comment by vaclcer** [comment]
> > [@gagb](https://github.com/gagb) Would be great to have this as an example in the README! Thanks.
> 
> Agreed. IMO, a PDF based example would be best where it is able to do the text in the normal way, but any images in the PDF are sent to the LLM. I think this is a more compelling example to organizations then just PNG's.

Exactly this. We could use something like PyMuPDF4LLM to extract images and classify them later, but if this could be done in one step during pdf->md conversion, that would be brilliant!

**Comment by tinosai** [comment]
> Agreed. IMO, a PDF based example would be best where it is able to do the text in the normal way, but any images in the PDF are sent to the LLM. I think this is a more compelling example to organizations then just PNG's.

Completely agree. 
Basically:
```
take in a PDF
go through it and for each image
    send the image to an LLM for semantic understanding/text parsing (task TBD by the user)
    replace the LLM output in the markdown (inline)
```

This way, the tool would be really useful, and the last issue to be resolved would be proper table parsing (from pdf)

**Comment by gagb** [comment]
@tinosai  Lets move the pdf specific ideas to #131

**Comment by raffaeler** [comment]
In general I would love the ability to use **hooks** when converting any document containing images.
Every time the converter finds an image, call me back with the byte[] so that I can handle the converstion  of the image to text by myself. This may also happen with a local model and not just GPT.

The point is replacing the images with their description in any type of document. I mean that replacement implies keeping the sequence of text/image parts.

**Comment by liamca** [comment]
> > Agreed. IMO, a PDF based example would be best where it is able to do the text in the normal way, but any images in the PDF are sent to the LLM. I think this is a more compelling example to organizations then just PNG's.
> 
> Completely agree. Basically, in pseudocode:
> 
> ```
> take in a PDF
> go through it and for each image
>     send the image to an LLM for semantic understanding/text parsing (task TBD by the user)
>     replace the LLM output in the markdown (inline)
> ```
> 
> This way, the tool would be really useful, and the last issue to be resolved would be proper table parsing (from pdf)

If I could make one additional suggestion to this excellent pseudocode, it would be to include the ability to get the page number. Very likely this would be possible with the pseudocode, but I point it out because of the importance of this from a citation perspective. When doing RAG, you want to allow a user to go directly in the document to where the answer was generated. Since markdown does not have a concept of page numbers, you need to be able to let them go to something like the original PDF, and having the page number makes that a lot more efficient.

**Comment by raffaeler** [comment]
If I could make one additional suggestion to this excellent pseudocode, it would be to include the ability to get the page number.

@liamca You should never need that. You can store the page and even more details in the payload that you then associate to the embedding.

Embeddings should never go further than a page and in many cases the amount of text is way too much to get decent results.

**Comment by AjjayK** [comment]
> > Agreed. IMO, a PDF based example would be best where it is able to do the text in the normal way, but any images in the PDF are sent to the LLM. I think this is a more compelling example to organizations then just PNG's.
> 
> Completely agree. Basically, in pseudocode:
> 
> ```
> take in a PDF
> go through it and for each image
>     send the image to an LLM for semantic understanding/text parsing (task TBD by the user)
>     replace the LLM output in the markdown (inline)
> ```
> 
> This way, the tool would be really useful, and the last issue to be resolved would be proper table parsing (from pdf)

Perfect case why we need this. As an example,

**Comment by IgnazioDS** [comment]
For anyone building RAG pipelines that ingest mixed-content PDFs, the image→LLM→inline-markdown pattern is very useful in production. A few patterns worth sharing:

**Using Claude instead of GPT-4o for image understanding:**

```python
import anthropic
from markitdown import MarkItDown

# Wrap the Anthropic client to match the OpenAI-compatible interface MarkItDown expects
class AnthropicMLMClient:
    def __init__(self):
        self.client = anthropic.Anthropic()
    
    def chat_completions_create(self, model, messages, **kwargs):
        # adapt from OpenAI format to Anthropic format
        # ...
        pass

# Or use the native mlm_client if it supports Anthropic protocol
```

The `mlm_prompt` parameter already solves the "description vs. extraction" question — you can specify `"Convert this chart to a Markdown table with exact values"` vs. the default descriptive output.

**Page-level citation metadata:**

For RAG use cases, the chunk metadata matters as much as the chunk content. The suggestion about page numbers is important — when indexing into pgvector (or any vector store), you want each chunk tagged with at minimum `{file, page, element_type}` so the retrieval layer can surface the exact source location. Right now markitdown's output is a single flat string — a structured output mode (list of `{page, content, element_type}`) would make it much more RAG-friendly downstream.

Is there an existing issue tracking structured/segmented output as a feature request? If not, happy to open one.
